### PR TITLE
Support for service clients

### PR DESCRIPTION
**Pull request by pablogs9**
This PR adds support for 
- service and client
- guard condition

 to the RCLC. It addresses issue  https://github.com/micro-ROS/rclc/issues/36

TODO: 
- [ ] Refactor all examples along micro-ROS repos:
- https://github.com/micro-ROS/micro-ROS-demos/pull/31
- FreeRTOS
- Zephyr
- Arduino

### CODE REVIEW

**Comment by pablogs9** [comment]
Hello @JanStaschulat, I'm opening this just to discuss the support of services in RCLC.

Maybe we need this in order to do a demo and I would like to ask you when do you plan to have a mergeable version and if we can help somehow to speed this up.

**Comment by JanStaschulat** [comment]
@pablogs9 My plan is to have a working version by the end of this week for service client support. In a second step then to support guard_conditions. With them you can execute callbacks as requested on F2F meeting.

**Comment by pablogs9** [comment]
Perfect @JanStaschulat, nice to hear this.

Do you want me to close this PR or you are ok with reusing it?

**Comment by JanStaschulat** [comment]
@pablogs9 We can reuse this PR. I will continue working on the branch feature/support_for_service_clients. If it works, we can review the code changes using this pull request.

**Comment by JanStaschulat** [comment]
@pablogs9 @ralph-lange Implementation and unit test done for service/client support. Give it a try.
See example in test_executor.cpp , unit test executor_test_service

File: rclc/include/rclc/client.h
Comment (pablogs9) [inline rclc/include/rclc/client.h]:
```suggestion
  const char * service_name);
```

File: rclc/include/rclc/client.h
Comment (pablogs9) [inline rclc/include/rclc/client.h]:
```suggestion
 * \param[in] service_name the name of service
```

File: rclc/include/rclc/client.h
Comment (pablogs9) [inline rclc/include/rclc/client.h]:
```suggestion
 * \param[in] service_name the name of client topic
```

File: rclc/include/rclc/client.h
Comment (pablogs9) [inline rclc/include/rclc/client.h]:
```suggestion
  const char * service_name);
```

File: rclc/include/rclc/executor_handle.h
Comment (pablogs9) [inline rclc/include/rclc/executor_handle.h]: These callbacks could be unified in anonymous unions because they are not going to coexists, what do you think @JanStaschulat ? Something like:

```c
union{
  rclc_callback_t callback;
  rclc_service_callback_t service_callback;
  rclc_client_callback_t client_callback;
}
```

File: rclc/include/rclc/service.h
Comment (pablogs9) [inline rclc/include/rclc/service.h]:
```suggestion
  const char * service_name);
```

File: rclc/include/rclc/service.h
Comment (pablogs9) [inline rclc/include/rclc/service.h]:
```suggestion
 * \param[in] service_name the name of service topic
```

File: rclc/include/rclc/service.h
Comment (pablogs9) [inline rclc/include/rclc/service.h]:
```suggestion
 * \param[in] service_name the name of service topic
```

File: rclc/include/rclc/service.h
Comment (pablogs9) [inline rclc/include/rclc/service.h]:
```suggestion
  const char * service_name);
```

File: rclc/src/rclc/client.c
Comment (pablogs9) [inline rclc/src/rclc/client.c]:
```suggestion
  const char * service_name)
```

File: rclc/src/rclc/client.c
Comment (pablogs9) [inline rclc/src/rclc/client.c]:
```suggestion
  const char * service_name)
```

File: rclc/include/rclc/client.h
Comment (ralph-lange) [inline rclc/include/rclc/client.h]:
```suggestion
// Copyright (c) 2020 - for information on the respective copyright owner
```

File: rclc/include/rclc/executor.h
Comment (ralph-lange) [inline rclc/include/rclc/executor.h]:
```suggestion
 * * An error is returned if {@link rclc_executor_t.handles} array is full.
```

File: rclc/include/rclc/executor.h
Comment (ralph-lange) [inline rclc/include/rclc/executor.h]:
```suggestion
 * * An error is returned if {@link rclc_executor_t.handles} array is full.
```

File: rclc/src/rclc/executor.c
Comment (pablogs9) [inline rclc/src/rclc/executor.c:723]: What if the user doesn't want to answer the service? Is this a possibility in the concept of service in ROS 2?

Maybe we can check if the response message is not null and only `rcl_send_response` if so.

Also, if the RCLC is going to handle the `rcl_send_response` why does the user needs the `req_id`?

File: rclc/include/rclc/service.h
Comment (ralph-lange) [inline rclc/include/rclc/service.h]:
```suggestion
// Copyright (c) 2020 - for information on the respective copyright owner
```

File: rclc/src/rclc/client.c
Comment (ralph-lange) [inline rclc/src/rclc/client.c]:
```suggestion
// Copyright (c) 2020 - for information on the respective copyright owner
```

File: rclc/src/rclc/service.c
Comment (pablogs9) [inline rclc/src/rclc/service.c]:
```suggestion
  const char * service_name)
```

File: rclc/src/rclc/service.c
Comment (pablogs9) [inline rclc/src/rclc/service.c]:
```suggestion
  const char * service_name)
```

File: rclc/src/rclc/service.c
Comment (pablogs9) [inline rclc/src/rclc/service.c]:
```suggestion
    service_name,
```

File: rclc/src/rclc/service.c
Comment (pablogs9) [inline rclc/src/rclc/service.c]:
```suggestion
    service_name,
```

File: rclc/src/rclc/service.c
Comment (pablogs9) [inline rclc/src/rclc/service.c]:
```suggestion
    service_name, "service_name is a null pointer", return RCL_RET_INVALID_ARGUMENT);
```

File: rclc/include/rclc/service.h
Comment (ralph-lange) [inline rclc/include/rclc/service.h]: TODO? See also next two lines.

File: rclc/src/rclc/executor.c
Comment (ralph-lange) [inline rclc/src/rclc/executor.c:297]: Also check response_msg and callback for NULL? At least, later, client_callback is called without checking for NULL, which will give a runtime error.

File: rclc/include/rclc/executor_handle.h
Comment (JanStaschulat) [reply rclc/include/rclc/executor_handle.h]: Yes. Resolved in be74fd4

File: rclc/src/rclc/service.c
Comment (ralph-lange) [inline rclc/src/rclc/service.c]:
```suggestion
// Copyright (c) 2020 - for information on the respective copyright owner
```

File: rclc/test/rclc/test_client.cpp
Comment (ralph-lange) [inline rclc/test/rclc/test_client.cpp]:
```suggestion
// Copyright (c) 2020 - for information on the respective copyright owner
```

File: rclc/include/rclc/executor_handle.h
Comment (ralph-lange) [inline rclc/include/rclc/executor_handle.h:36]: Are guard conditions fully implemented by this PR?

**Review by ralph-lange — CHANGES_REQUESTED** [review]

File: rclc/src/rclc/executor.c
Comment (JanStaschulat) [reply rclc/src/rclc/executor.c:723]: There is a design choice for the service_callback:
1. **service_callback(const void * req_msg, rmw_request_id_t * id, void * resp_msg)**
the service_callback gets as parameter the request_msg and (possibly) assigns the response_msg

2. **service_callback(const void * req_msg, rmw_request_id_t * id, rcl_service_t * service)** 
the service_callback also gets a pointer to the service. Then whether to send a response or not is up to the service_callback. 

3. **service_callback(const void * req_msg, void * resp_msg)**
The service callback gets only a pointer to the request message and pointer to response_message.

I decided for option 1. I thought only the executor should know about the service object and call the rcl_send_response function. The user shall only compute the answer. But I am open for other options that would make more sense - or be easier for the user.

Regarding reg_id (option 3) 
I was not sure, if the user might want to know about it. Maybe the application also wants to know about which sequence numbe is processed - if one is maybe lost? Do you think the req_id is only for internal administration?

File: rclc/src/rclc/executor.c
Comment (JanStaschulat) [reply rclc/src/rclc/executor.c:297]: Good catch. resolved.
updated rclc_executor_add_client and rclc_executor_add_service functions as well as the unit tests in  8e95e92

**Review by JanStaschulat — CHANGES_REQUESTED** [review]
Discuss about 
- service_callback signature
- implementation of guard conditions in this PR => done

**Comment by JanStaschulat** [comment]
@pablogs9 Took your "addtwoints" service/client example from [micro-ROS-demos](https://github.com/micro-ROS/micro-ROS-demos/tree/feature/update_services_examples/rclc) and added them to [rclc/rclc_examples](https://github.com/micro-ROS/rclc/tree/feature/support_for_service_clients/rclc_examples) with minor changes in commit 97c4fd6.

The service/client-example works, however  sometimes sporadic initialization errors occur. All tests done with ROS 2 Foxy on Linux PC. Do you get similar errors on FreeRTOS/Zephyr/NuttX? => resolved in 9b77c60

**CLIENT NODE**
jst3si@RNGX5596:**ros2 run rclc_examples example_client_node** 
[ERROR] [1605013436.369456148] [rclc]: [rclc_client_init_default] Error in rcl_client_init: client already initialized, or memory was unintialized, at /home/jst3si/ros2_try/src/ros2/rcl/rcl/src/rcl/client.c:74

Failed status on line 51: 100. Aborting.
jst3si@RNGX5596: ros2 run rclc_examples example_client_node 
Send service request 24 + 42. Seq 1
INFO: rcl_wait timeout 10 ms
Received service response 24 + 42 = 66. Seq 1


**SERVICE_NODE**
jst3si@RNGX5596:**ros2 run rclc_examples example_service_node** 
[ERROR] [1605013372.887516121] [rclc]: [rclc_service_init_default] Error in rcl_service_init: service already initialized, or memory was unintialized, at /home/jst3si/ros2_try/src/ros2/rcl/rcl/src/rcl/service.c:71

Failed status on line 57: 100. Aborting.
jst3si@RNGX5596: ros2 run rclc_examples example_service_node 
INFO: rcl_wait timeout 10 ms
Service request value: 24 + 42. Seq 1

**Comment by JanStaschulat** [comment]
@pablogs9 Bug of uninitialized memory resolved in 9b77c60. You need to initialize e.g. a client with: 
    rcl_client_t client = rcl_get_zero_initialized_client();

**Comment by pablogs9** [comment]
Regarding the discussion on if the executor always should answer the service request:

As the executor can't see if the user has set or not any data in the response the only way I see is to pass a pointer to pointer and let the user set it to NULL if he/she don't want to answer the service request. Something like:

```c
case SERVICE:
        void ** tmp_pointer = (void *) &handle->data_response_msg;
        handle->service_callback(handle->data, tmp_pointer);
// If callback wants an answer should do: mytype * msg_out = (* mytype) *tmp_pointer; 
// If not: *tmp_pointer = NULL; 
        if(*tmp_pointer != NULL){
                rc = rcl_send_response(handle->service, &handle->req_id, tmp_pointer);
        }
```

Anyhow, I am not sure I like this solution. If you don't like it, maybe we can keep the always-answer policy by now.
If the executor is in charge of answering the service request I don't see any need of passing `handle->req_id` to the user.

Thought @JanStaschulat @jamoralp ?

**Comment by JanStaschulat** [comment]
@jamoralp  @ralph-lange @pablogs9 
regarding service_callback signature: I looked in rclcpp implementation:
- service callback gets as parameter a pointer to request_msg and response_msg. But not a pointer to rcl_service_t. two options: 
  - A) a service callback has as parameter only pointer to request_msg and response_msg
  - B) a service callback has as parameter additionally the request_header (aka rmw_request_id_t) also
- service_callback is called by Executor
- service_response (aka rcl_send_response()) is called by Executor irrespective whether response_msg was updated by service callback function


here are the links to the relevant code snippets:
[Executor::execute_service](https://github.com/ros2/rclcpp/blob/6820dac31535a2c37479037594210cd832d6cf82/rclcpp/src/rclcpp/executor.cpp#L439) is the function called from the Executor to process a service request:
```C
void
Executor::execute_service(rclcpp::ServiceBase::SharedPtr service)
{
  auto request_header = service->create_request_header();
  std::shared_ptr<void> request = service->create_request();
  take_and_do_error_handling(
    "taking a service server request from service",
    service->get_service_name(),
    [&]() {return service->take_type_erased_request(request.get(), *request_header);},
    [&]() {service->handle_request(request_header, request);});
}
```

After fetching the request message, it calls [service->handle_request](https://github.com/ros2/rclcpp/blob/6820dac31535a2c37479037594210cd832d6cf82/rclcpp/include/rclcpp/service.hpp#L342)
```C
  void
  handle_request(
    std::shared_ptr<rmw_request_id_t> request_header,
    std::shared_ptr<void> request) override
  {
    auto typed_request = std::static_pointer_cast<typename ServiceT::Request>(request);
    auto response = std::make_shared<typename ServiceT::Response>();
    any_callback_.dispatch(request_header, typed_request, response);
    send_response(*request_header, *response);
  }
```
Which calls the [dispatch](https://github.com/ros2/rclcpp/blob/6820dac31535a2c37479037594210cd832d6cf82/rclcpp/include/rclcpp/any_service_callback.hpp#L86) function of any_callback_, which is of type AnyServiceCallback:
```C
  void dispatch(
    std::shared_ptr<rmw_request_id_t> request_header,
    std::shared_ptr<typename ServiceT::Request> request,
    std::shared_ptr<typename ServiceT::Response> response)
  {
    TRACEPOINT(callback_start, (const void *)this, false);
    if (shared_ptr_callback_ != nullptr) {
      (void)request_header;
      shared_ptr_callback_(request, response);
    } else if (shared_ptr_with_request_header_callback_ != nullptr) {
      shared_ptr_with_request_header_callback_(request_header, request, response);
    } else {
      throw std::runtime_error("unexpected request without any callback set");
    }
    TRACEPOINT(callback_end, (const void *)this);
  }
```
This is interesting: The class AnyServiceCallback can be instantiated either with a callback that takes only two arguments (request message, response message) or three arguments (request_header, request message, response message). This is checked by comparing shared_ptr_callback_ or shared_ptr_with_request_header_callback_ with NULL pointer. (see src code below). Depending which callback is used, one of the following functions is called:
```C
shared_ptr_callback_(request, response);
shared_ptr_with_request_header_callback_(request_header, request, response);
```

The assignment of the ptr_callbacks is implemented in [templated ::set function](https://github.com/ros2/rclcpp/blob/6820dac31535a2c37479037594210cd832d6cf82/rclcpp/include/rclcpp/any_service_callback.hpp#L58)
```C
  template<
    typename CallbackT,
    typename std::enable_if<
      rclcpp::function_traits::same_arguments<
        CallbackT,
        SharedPtrCallback
      >::value
    >::type * = nullptr
  >
  void set(CallbackT callback)
  {
    shared_ptr_callback_ = callback;
  }

  template<
    typename CallbackT,
    typename std::enable_if<
      rclcpp::function_traits::same_arguments<
        CallbackT,
        SharedPtrWithRequestHeaderCallback
      >::value
    >::type * = nullptr
  >
  void set(CallbackT callback)
  {
    shared_ptr_with_request_header_callback_ = callback;
  }
```

After the service callback is called, we are back in the handle_request() function, which calls 
[send_response(*request_header, *response)](https://github.com/ros2/rclcpp/blob/6820dac31535a2c37479037594210cd832d6cf82/rclcpp/include/rclcpp/service.hpp#L361) 
```C
  void
  send_response(rmw_request_id_t & req_id, typename ServiceT::Response & response)
  {
    rcl_ret_t ret = rcl_send_response(get_service_handle().get(), &req_id, &response);

    if (ret != RCL_RET_OK) {
      rclcpp::exceptions::throw_from_rcl_error(ret, "failed to send response");
    }
  }
```
... which calls rcl_send_response() without checking the contents of the response message.

Based on this, I propose: 
- give the user the choice: service_callback with or without request_header as parameter
- the executor calls that function accordingly with or without the request_header
- the executor sends the response message (blindly) irrespective whether the response message has been updated or not by the service callback

That is the same behavior as in rclcpp.

File: rclc/include/rclc/executor_handle.h
Comment (JanStaschulat) [reply rclc/include/rclc/executor_handle.h:36]: yes: 1ce5446

File: rclc/src/rclc/executor.c
Comment (JanStaschulat) [reply rclc/src/rclc/executor.c:723]: See my comment: https://github.com/micro-ROS/rclc/pull/44#issuecomment-724959085

**Comment by pablogs9** [comment]
I agree with your proposal.

Another small question regarding  9b77c60: why the `rcl_get_zero_initialized_client()` is not call inside `rclc_client_init_default()`?

File: rclc/include/rclc/executor_handle.h
Comment (jamoralp) [inline rclc/include/rclc/executor_handle.h]:
```suggestion
/// Type definition for client callback function
```

**Comment by JanStaschulat** [comment]
@pablogs9 
> Another small question regarding 9b77c60: why the rcl_get_zero_initialized_client() is not call inside rclc_client_init_default()?

All init functions of timer, subscription, publisher, service,client and executor assume that the handle is already zero_initialized. This is the API of the rcl_init_* function. So I did it the same way for rclc_init_* functions.

But I agree, that fewer number of calls - the easier it is for the user. I'll have a look in all the rclc_init_functions. If it works, I would create a new PR to address this issue, as it is a change in the API and we as well as everyone else, who is using rclc_executor, will have to update his/her application as well.

As service/client is a new feature, calling get_zero_initialized_*() could be in the rclc_init_* function.

**Comment by pablogs9** [comment]
Why is this changing the API? It is just that the `rcl_publisher_t` passed as the first argument of `rclc_publisher_init_default` is going to be set to zero inside the function. Where is the API change? In the worst case, the object is going to be set to zero twice.

**Review by JanStaschulat — DISMISSED** [review]
@pablogs9 I agree. Implemented as suggested in comment https://github.com/micro-ROS/rclc/pull/44#issuecomment-724959085. Please re-review changes:

- updated service_callback/client_callback signature, so the user can decide:
   - (A) without request_id
   - (B) with request_id
- updated init functions for client and service to include rcl_get_zero_initialized_client() and rcl_get_zero_initialized_service(). Benefit: the used does not have to call get_zero_initialized_x function.

**Comment by JanStaschulat** [comment]
@ralph-lange As far as I can see, I addressed all your review requests. Please re-review.

**Review by JanStaschulat — APPROVED** [review]

**Comment by pablogs9** [comment]
@JanStaschulat nice, I'm completely ok with having the same approach as the RCLCPP. I can't approve this PR bc I created it, so when you all approve it we can merge.

**Comment by JanStaschulat** [comment]
@pablogs9 thanks for your https://github.com/micro-ROS/rclc/pull/44#issuecomment-725359388
See PR https://github.com/micro-ROS/rclc/pull/58 , which simplifies initialization of other objects (publisher, subscription, node, executor, timer) too. i.e: without calling the function rcl_get_zero_initialized_x() for those objects. 

You might want to update your examples too after PR client/service and PR simplify init function are merged.

**Review by ralph-lange — APPROVED** [review]

**Pull request merged**
